packages/ovault-evm/contracts/OVaultComposer.sol

```diff
@@ -126,116 +152,197 @@ contract OVaultComposer is IOVaultComposer, ReentrancyGuard {
+        if (_failedGuidState(failedMessage) != FailedState.CanOnlyRetry) revert CanNotRetry(_guid);
 
         SendParam memory sendParam = failedMessage.sendParam;
+        uint256 prePaidMsgValue = failedMessage.msgValue;
 
-        sendParam.extraOptions = _extraOptions;
+        if (removeExtraOptions) {
+            (bool sent, ) = payable(REFUND_OVERPAY_ADDRESS).call{ value: prePaidMsgValue }("");
+            require(sent, "Failed to send Ether");
 
-        delete failedMessages[_guid];
-        _send(failedMessage.oft, sendParam);
+            sendParam.extraOptions = "";
+            prePaidMsgValue = 0;
+        }
+
+        this.sendFailedMessage{ value: msg.value }(_guid, failedMessage.oft, sendParam, prePaidMsgValue, tx.origin);
         emit Retried(_guid, failedMessage.oft);
     }
 
-    /// @dev Retry mechanism for transactions that failed due to slippage. This can revert.
-    /// @dev Failure case when there is a LayerZero config issue - ex: dvn config
-    function retryWithSwap(bytes32 _guid, bytes calldata _extraOptions) external payable {
+    /// @dev Performs the ovault action for a failed swap. If we fail on the send then the GUID enters the retry state for manual execution.
+    function retryWithSwap(bytes32 _guid, bool skipRetry) external payable nonReentrant {
         FailedMessage memory failedMessage = failedMessages[_guid];
-        if (failedGuidState(_guid) != FailedState.CanRetryWithSwap) revert CanNotRetry(_guid);
-
-        SendParam memory sendParam = failedMessage.sendParam;
-        sendParam.extraOptions = _extraOptions;
+        if (_failedGuidState(failedMessage) != FailedState.CanRefundOrRetryWithSwap) revert CanNotSwap(_guid);
+
+        /// @dev Disable refund. If this call reverts refund is still possible.
+        failedMessages[_guid].refundOFT = address(0);
+
+        uint256 srcAmount = failedMessage.refundSendParam.amountLD;
+
+        uint256 vaultAmount = _executeOVaultActionWithSlippageCheck(
+            failedMessage.refundOFT,
+            srcAmount,
+            failedMessage.sendParam.minAmountLD
+        );
+
+        failedMessage.sendParam.amountLD = vaultAmount;
+        emit SwappedTokens(_guid);
+
+        /// @dev Regardless of the outcome of skipRetry and sendFailedMessage, the failedMessage is now in a RETRY only state.
+        /// @dev try..catch to accumulate the msg.value in the case of a failed send
+        if (!skipRetry) {
+            try
+                this.sendFailedMessage{ value: msg.value }(
+                    _guid,
+                    failedMessage.oft,
+                    failedMessage.sendParam,
+                    failedMessage.msgValue,
+                    tx.origin
+                )
+            {
+                emit Retried(_guid, failedMessage.oft);
+            } catch {
+                failedMessages[_guid].msgValue += msg.value;
+                emit SendFailed(_guid, failedMessage.oft);
+            }
```

Comment:
when `skipRetry == true` something should also be done with msg.value:
either revert or add it to `failedMessages[_guid].msgValue`